internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,570 @@
+	if rbacregistry.EscalationAllowed(ctx) {
+		return nil
+	}
+
+	roleRef := rbacinternal.RoleRef{}
+	err := rbacv1helpers.Convert_v1_RoleRef_To_rbac_RoleRef(&clusterRoleBinding.RoleRef, &roleRef, nil)
+	if err != nil {
+		return err
+	}
+
+	if rbacregistry.BindingAuthorized(ctx, roleRef, metav1.NamespaceNone, ec.authorizer) {
+		return nil
+	}
+
+	rules, err := ec.ruleResolver.GetRoleReferenceRules(ctx, clusterRoleBinding.RoleRef, metav1.NamespaceNone)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if clusterRoleBinding.RoleRef.Kind == "ClusterRole" {
+		if manifestClusterRole, ok := ec.extraClusterRoles[types.NamespacedName{Name: clusterRoleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestClusterRole.Rules...)
+		}
+	}
+
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ec *escalationChecker) checkRoleEscalation(ctx context.Context, role *rbacv1.Role) error {
+	if rbacregistry.EscalationAllowed(ctx) || rbacregistry.RoleEscalationAuthorized(ctx, ec.authorizer) {
+		return nil
+	}
+
+	rules := role.Rules
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ec *escalationChecker) checkRoleBindingEscalation(ctx context.Context, roleBinding *rbacv1.RoleBinding) error {
+	if rbacregistry.EscalationAllowed(ctx) {
+		return nil
+	}
+
+	roleRef := rbacinternal.RoleRef{}
+	err := rbacv1helpers.Convert_v1_RoleRef_To_rbac_RoleRef(&roleBinding.RoleRef, &roleRef, nil)
+	if err != nil {
+		return err
+	}
+	if rbacregistry.BindingAuthorized(ctx, roleRef, roleBinding.Namespace, ec.authorizer) {
+		return nil
+	}
+
+	rules, err := ec.ruleResolver.GetRoleReferenceRules(ctx, roleBinding.RoleRef, roleBinding.Namespace)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	switch roleRef.Kind {
+	case "ClusterRole":
+		if manifestClusterRole, ok := ec.extraClusterRoles[types.NamespacedName{Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestClusterRole.Rules...)
+		}
+	case "Role":
+		if manifestRole, ok := ec.extraRoles[types.NamespacedName{Namespace: roleBinding.Namespace, Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestRole.Rules...)
+		}
+	}
+
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+var fullAuthority = []rbacv1.PolicyRule{
+	{Verbs: []string{"*"}, APIGroups: []string{"*"}, Resources: []string{"*"}},
+	{Verbs: []string{"*"}, NonResourceURLs: []string{"*"}},
+}
+
+func parseEscalationErrorForMissingRules(ecError error) ([]rbacv1.PolicyRule, string) {
+	// Regex to capture namespace and serviceaccount
+	userRegex := regexp.MustCompile(`system:serviceaccount:(?P<Namespace>[^:]+):(?P<ServiceAccount>[^"]+)`)
```

Comment:
This is fragile because it assumes the user will be a service account. I don't think we should make that assumption:
1. We've got a feature coming soon around synthetic auth
2. It would make this code less reusable in CLI contexts where the user is not in our control.

I think we already know the namespace when we call this function though (it's the namespace of the object where doing the escalation check against).

Reply:
Here's the code that generates the string. I think it would probably be best if we wrote the regex that could fully describe (and extract the relevant pieces) of the full string.

https://github.com/kubernetes/kubernetes/blob/3d0594d4d92653fcefea0b91363f473ae457c6b8/pkg/registry/rbac/validation/rule.go#L81-L84

Reply:
(that would also help with extracting out the rule resolution errors that I mentioned above)

Reply:
This thread is outdated because I updated the regexp. @joelanford have you had a chance to look at the new one?

Reply:
think this is all settled from later commits. resolving